### PR TITLE
Add hermes publisher to DC Asset, Virtualserver and Cluster

### PR DESCRIPTION
**Pull request by ghost**
Add option for sending data to django-powerdns-dnssec app.
Data includes information to create auto txt records.

### CODE REVIEW

File: src/ralph/assets/publishers.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
+# -*- coding: utf-8 -*-
+import pyhermes
+from django.conf import settings
+from django.db.models.signals import post_save
+from django.dispatch import receiver
+
+from ralph.data_center.models.physical import DataCenterAsset
```
Comment (mkurek) [inline src/ralph/assets/publishers.py]: why not put it in dns package?

File: src/ralph/data_center/models/physical.py [hunk 2]
```diff
@@ -410,6 +411,55 @@ def save(self, *args, **kwargs):
...
+                self.rack.name
+            ])
+        if self.position:
+            location.append(str(self.position))
+        if self.slot_no:
+            location.append(str(self.slot_no))
+        return location
+
+    @property
+    def publish_data(self):
```
Comment (mkurek) [inline src/ralph/data_center/models/physical.py]: consider moving it to `NetworkableBaseObject`

File: src/ralph/data_center/models/physical.py [hunk 3]
```diff
@@ -410,6 +411,55 @@ def save(self, *args, **kwargs):
...
+                self.rack.server_room.name,
+                self.rack.name
+            ])
+        if self.position:
+            location.append(str(self.position))
+        if self.slot_no:
+            location.append(str(self.slot_no))
+        return location
+
+    @property
```
Comment (mkurek) [inline src/ralph/data_center/models/physical.py:446]: why not function?

File: src/ralph/helpers.py [hunk 4]
```diff
@@ -9,6 +9,27 @@
...
+    publish_data = []
+    for purpose, content in data:
+        data = {
+            'name': hostname,
+            'target_owner': target_owner,
+            'owner': owner,
+            'purpose': purpose,
+            'content': content,
+        }
+    return publish_data
```
Comment (mkurek) [inline src/ralph/helpers.py]: I don't get this function. What's the purpose of it? And `publish_data` will always be `[]`

File: src/ralph/assets/publishers.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+from ralph.data_center.models.virtual import Cluster
+from ralph.virtual.models import VirtualServer
+
+
+#TODO:: change topic name, shouldn't this be full path or from settings?
+@pyhermes.publisher(topic='auto_txt_record', auto_publish_result=True)
+def publish_data_to_dnsaaas(obj):
+    return obj.publish_data
```
Comment (mkurek) [inline src/ralph/assets/publishers.py]: check here if dnsaas integration is enabled

Comment (ghost) [inline src/ralph/assets/publishers.py]: https://github.com/allegro/ralph/blob/ng/src/ralph/settings/base.py#L349
is this enough? or you meant only the publisher registration?

File: src/ralph/assets/utils.py [hunk 6]
```diff
@@ -1,7 +1,34 @@
...
+from django.conf import settings
 from django.db import transaction
 
 
+class DNSaaSPublisherMixin:
+    """Generate data formatted for DNSaaS auto txt update"""
+    def get_auto_txt_data(self):
+        data = []
+        for purpose_name, content in  (
+            ('class_name', self.configuration_path.class_name if self.configuration_path else ''),  # noqa
```
Comment (mkurek) [inline src/ralph/assets/utils.py:11]: what about service_env?

File: src/ralph/assets/utils.py [hunk 7]
```diff
@@ -1,7 +1,34 @@
...
+            ('module_name', self.configuration_path.module.name if self.configuration_path else ''),  # noqa
```
Comment (mkurek) [inline src/ralph/assets/utils.py:12]: please add another entry `configuration_path` pointing to `self.configuration_path.path`

File: src/ralph/assets/utils.py [hunk 8]
```diff
@@ -1,7 +1,34 @@
...
+                continue
+            data.append({
+                'name': self.hostname or '',
+                'purpose': purpose,
+                'content': content,
+            })
+        return data
+
+
+
```
Comment (mkurek) [inline src/ralph/assets/utils.py]: why so many new lines?

File: src/ralph/data_center/models/virtual.py [hunk 9]
```diff
@@ -77,6 +79,13 @@ class Cluster(
     def __str__(self):
         return '{} ({})'.format(self.name or self.hostname, self.type)
 
+    def get_location(self):
+        return self.masters[0].get_location() if self.masters else None
+
+    @property
+    def model(self):
+        return self.masters[0].model if self.masters else None
```
Comment (mkurek) [inline src/ralph/data_center/models/virtual.py]: use `self.type.name` here

File: src/ralph/assets/utils.py [hunk 10]
```diff
@@ -1,7 +1,33 @@
...
 
 
+class DNSaaSPublisherMixin:
+    """Generate data formatted for DNSaaS auto txt update"""
+    def get_auto_txt_data(self):
+        data = []
+        for purpose_name, content in (
+            ('class_name', self.configuration_path.class_name if self.configuration_path else ''),  # noqa
+            ('module_name', self.configuration_path.module.name if self.configuration_path else ''),  # noqa
+            ('path', self.configuration_path.path if self.configuration_path else ''),  # noqa
```
Comment (mkurek) [inline src/ralph/assets/utils.py]: path -> configuration_path

File: src/ralph/assets/utils.py [hunk 11]
```diff
@@ -1,7 +1,33 @@
...
+            ('service_env', str(self.service_env) if self.service_env else ''),  # noqa
+            ('model', self.model.name if self.model else ''),
```
Comment (mkurek) [inline src/ralph/assets/utils.py]: use `str(self.model)` instead of model name

File: src/ralph/dns/publishers.py [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+from ralph.data_center.models.physical import DataCenterAsset
+from ralph.data_center.models.virtual import Cluster
+from ralph.virtual.models import VirtualServer
+
+
+def _publish_data_to_dnsaaas(obj):
+    publish_data = []
+    for data in obj.get_auto_txt_data():
+        data['owner'] = (
+            get_current_user().username if get_current_user() else ''
```
Comment (mkurek) [inline src/ralph/dns/publishers.py]: don't call `get_current_user` twice

File: src/ralph/settings/base.py [hunk 13]
```diff
@@ -344,7 +344,32 @@ def os_env_true(var, default=''):
...
+)
+# define names of values send to DNSAAS for:
+# DataCenterAsset, Cluster, VirtualServer
+DNSAAS_AUTO_TXT_RECORD_PURPOSE_MAP = {
+    # self.configuration_path.class will be send as 'VENTURE'
+    'class_name': 'VENTURE',
+    # self.configuration_path.module will be send as 'ROLE'
+    'module_name': 'ROLE',
+    # self.configuration_path.path will be send as 'PATH'
+    'path': 'PATH',
```
Comment (mkurek) [inline src/ralph/settings/base.py]: configuration path

File: src/ralph/assets/utils.py (hunk 10 above)
Comment (ghost) [inline src/ralph/assets/utils.py]: I'll also change the purpose value to `CONFIGURATION_PATH`, ok?

**Comment by mkurek** [comment]
:ok:

**Comment by ar4s** [comment]
:ok:

File: src/ralph/virtual/models.py [hunk 14]
```diff
@@ -258,6 +264,13 @@ class VirtualServer(AdminAbsoluteUrlMixin, NetworkableBaseObject, BaseObject):
     # TODO: remove this field
     cluster = models.ForeignKey(Cluster, blank=True, null=True)
 
+    def get_location(self):
+        return self.parent.get_location() if self.parent else None
```
Comment (mkurek) [inline src/ralph/virtual/models.py]: please add to the location hostname of the parent

Comment (ghost) [inline src/ralph/virtual/models.py]: I don't get it ;P
Example?

Comment (mkurek) [inline src/ralph/virtual/models.py]: for virtual server, its location should be: `DC1 / SR1 / Rack 101 / 10 / s12345.mydc.net` - the last part is hostname of its hypervisor.

**Comment by mkurek** [comment]
:ok:

**Pull request merged**